File: codes/leetcode/BestTimetoBuyandSellStockwithCooldown.cpp
/*
************************************************
username    :   smmehrab
fullname    :   s.m.mehrabul islam
email       :   mehrab.24csedu.001@gmail.com
institute   :   university of dhaka, bangladesh
session     :   2017-2018
************************************************
*/
/* 
    ** 3 States **

    (1)
    a -> hold -> a
      -> buy -> b

    (2) 
    b -> hold -> b
      -> sell -> c
    
    (3)
    c -> hold -> a

*/
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int a=0, b=INT_MIN, c=0, previousA;
        for(int price : prices){
            previousA = a;
            a = max(a, c);
            c = b + price;
            b = max(b, previousA-price);
        }
        return max(a, c);
    }
};

File: codes/leetcode/BestTimetoBuyandSellStockwithCooldown.py
# ************************************************
# username  :   smmehrab
# fullname  :   s.m.mehrabul islam
# email     :   smmehrabul-2017614964@cs.du.ac.bd
# institute :   university of dhaka, bangladesh
# reg       :   2017614964
# ************************************************

class Solution:

    def _dfs(self, i, action):
        # Base
        if i>= self.n:
            return 0
        if (i, action) in self.dp:
            return self.dp[(i, action)]
        # Cooldown
        cooldown = self._dfs(i+1, action)
        # Selling
        if action == 1:
            sell = self._dfs(i+2, 1-action) + self.prices[i]
            self.dp[(i, action)] = max(sell, cooldown)
        # Buying
        else:
            buy = self._dfs(i+1, 1-action) - self.prices[i]
            self.dp[(i, action)] = max(buy, cooldown)
        return self.dp[(i, action)]

    def maxProfit(self, prices: list[int]) -> int:
        self.n = len(prices)
        self.prices = prices
        self.dp = {}
        return self._dfs(0, 0)


